File: lang/typescript/cxlib/examples/transform.ts
/**
 * CX transform examples — demonstrates the TypeScript cxlib wrapper around libcx.
 *
 * Run from repo root:
 *   cd typescript/cxlib && npm run example
 */
import * as cx from '../src/index';
import * as fs from 'fs';
import * as path from 'path';

const EXAMPLES = path.resolve(__dirname, '..', '..', '..', '..', 'examples');

function read(name: string): string {
  return fs.readFileSync(path.join(EXAMPLES, name), 'utf8');
}

function section(title: string): void {
  console.log(`\n${'─'.repeat(60)}`);
  console.log(`  ${title}`);
  console.log('─'.repeat(60));
}

// ── article.cx ────────────────────────────────────────────────────────────────

section('article.cx  (source)');
console.log(read('article.cx'));

section('article.cx  →  CX  (canonical round-trip)');
console.log(cx.toCx(read('article.cx')));

section('article.cx  →  XML');
console.log(cx.toXml(read('article.cx')));

section("article.cx  →  JSON  (mixed content uses '_' for text runs)");
console.log(cx.toJson(read('article.cx')));

section('article.cx  →  YAML');
console.log(cx.toYaml(read('article.cx')));

// ── env.cx ────────────────────────────────────────────────────────────────────

section('env.cx  (source)');
console.log(read('env.cx'));

section('env.cx  →  CX  (canonical round-trip)');
console.log(cx.toCx(read('env.cx')));

section('env.cx  →  XML  (anchor/merge as cx:anchor / cx:merge attrs)');
console.log(cx.toXml(read('env.cx')));

section('env.cx  →  JSON');
console.log(cx.toJson(read('env.cx')));

section('env.cx  →  YAML');
console.log(cx.toYaml(read('env.cx')));

section('env.cx  →  TOML');
console.log(cx.toToml(read('env.cx')));

// ── config.cx ─────────────────────────────────────────────────────────────────

section('config.cx  →  JSON');
console.log(cx.toJson(read('config.cx')));

section('config.cx  →  YAML');
console.log(cx.toYaml(read('config.cx')));

section('config.cx  →  TOML');
console.log(cx.toToml(read('config.cx')));

// ── books.cx ──────────────────────────────────────────────────────────────────

section('books.cx  →  XML');
console.log(cx.toXml(read('books.cx')));

section('books.cx  →  JSON  (repeated elements auto-collect into arrays)');
console.log(cx.toJson(read('books.cx')));

// ── cross-format round-trips ──────────────────────────────────────────────────

section('books.xml   →  CX');
console.log(cx.xmlToCx(read('books.xml')));

section('books.json  →  CX');
console.log(cx.jsonToCx(read('books.json')));

section('config.yaml  →  CX');
console.log(cx.yamlToCx(read('config.yaml')));

section('config.toml  →  CX');
console.log(cx.tomlToCx(read('config.toml')));

// ── vcore.cx ──────────────────────────────────────────────────────────────────

section('vcore.cx  (source)');
console.log(read('vcore.cx'));

section('vcore.cx  →  CX  (short aliases expand; auto-array gains explicit :type[])');
console.log(cx.toCx(read('vcore.cx')));

section('vcore.cx  →  JSON  (auto-array, :[], triple-quoted, block content)');
console.log(cx.toJson(read('vcore.cx')));

section('vcore.cx  →  XML  (cx:type annotations; cx:block for block content)');
console.log(cx.toXml(read('vcore.cx')));

// ── doc.cx ────────────────────────────────────────────────────────────────────

section('doc.cx  (source)');
console.log(read('doc.cx'));

section('doc.cx  →  Markdown');
console.log(cx.toMd(read('doc.cx')));

section('doc.cx  →  XML');
console.log(cx.toXml(read('doc.cx')));

// ── doc.md ────────────────────────────────────────────────────────────────────

section('doc.md  (source)');
console.log(read('doc.md'));

section('doc.md  →  CX');
console.log(cx.mdToCx(read('doc.md')));

// ── AST inspection ────────────────────────────────────────────────────────────

// ── chapter.cx: XML-style structured document ────────────────────────────────

section('chapter.cx  (source)');
console.log(read('chapter.cx'));

section('chapter.cx  →  CX  (canonical)');
console.log(cx.toCx(read('chapter.cx')));

section('chapter.cx  →  XML  (structured document with sections and table)');
console.log(cx.toXml(read('chapter.cx')));

section('chapter.cx  →  JSON  (nested sections as nested objects)');
console.log(cx.toJson(read('chapter.cx')));

// ── post.cx: Markdown-style blog post ────────────────────────────────────────

section('post.cx  (source)');
console.log(read('post.cx'));

section('post.cx  →  Markdown  (CX markdown dialect → rendered Markdown)');
console.log(cx.toMd(read('post.cx')));

section('post.cx  →  CX  (canonical)');
console.log(cx.toCx(read('post.cx')));

section("article.cx  →  AST  →  CX  (round-trip: AST JSON back to canonical CX)");
console.log(cx.astToCx(cx.toAst(read('article.cx'))));

section("article.cx  →  CX  (compact)");
console.log(cx.toCxCompact(read('article.cx')));

// ── Document API ──────────────────────────────────────────────────────────────

const svcSrc = `[services
  [service name=auth  port=8080 active=true]
  [service name=api   port=9000 active=false]
  [service name=web   port=80   active=true]
]`;

section('Document API: CXPath select');
const svcDoc = cx.parse(svcSrc);
const first = svcDoc.select('//service');
console.log(`first service: ${first?.attr('name')}`);
for (const svc of svcDoc.selectAll('//service[@active=true]')) {
  console.log(`active: ${svc.attr('name')}`);
}

section('Document API: transform (immutable update)');
const updated = svcDoc.transform('services/service', el => { el.setAttr('name', 'renamed-auth'); return el; });
console.log(updated.toCx());

section('Document API: transform_all');
const allActive = svcDoc.transformAll('//service', el => { el.setAttr('active', true); return el; });
console.log(`active services after transform_all: ${allActive.selectAll('//service[@active=true]').length}`);

File: lang/ruby/cxlib/examples/transform.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# CX transform examples — demonstrates the Ruby cxlib wrapper around libcx.
#
# Run from the repo root:
#   /opt/homebrew/opt/ruby/bin/ruby ruby/cxlib/examples/transform.rb
#
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'cxlib'

EXAMPLES = File.expand_path('../../../../../examples', __FILE__)

def read(name)
  File.read(File.join(EXAMPLES, name))
end

def section(title)
  puts
  puts '#{'─' * 60}'
  puts "  #{title}"
  puts '#{'─' * 60}'
end

# ── article.cx: comments, mixed content, raw text, entity refs ───────────────

section 'article.cx  (source)'
puts read('article.cx')

section 'article.cx  →  CX  (canonical round-trip)'
puts CXLib.to_cx(read('article.cx'))

section 'article.cx  →  XML'
puts CXLib.to_xml(read('article.cx'))

section "article.cx  →  JSON  (mixed content uses '_' for text runs)"
puts CXLib.to_json(read('article.cx'))

section 'article.cx  →  YAML'
puts CXLib.to_yaml(read('article.cx'))

# ── env.cx: anchors, merges, comments ────────────────────────────────────────

section 'env.cx  (source)'
puts read('env.cx')

section 'env.cx  →  CX  (canonical round-trip)'
puts CXLib.to_cx(read('env.cx'))

section 'env.cx  →  XML  (anchor/merge as cx:anchor / cx:merge attrs)'
puts CXLib.to_xml(read('env.cx'))

section 'env.cx  →  JSON'
puts CXLib.to_json(read('env.cx'))

section 'env.cx  →  YAML'
puts CXLib.to_yaml(read('env.cx'))

section 'env.cx  →  TOML'
puts CXLib.to_toml(read('env.cx'))

# ── config.cx: typed scalars, arrays ─────────────────────────────────────────

section 'config.cx  →  JSON'
puts CXLib.to_json(read('config.cx'))

section 'config.cx  →  YAML'
puts CXLib.to_yaml(read('config.cx'))

section 'config.cx  →  TOML'
puts CXLib.to_toml(read('config.cx'))

# ── books.cx: repeated elements become arrays ─────────────────────────────────

section 'books.cx  →  XML'
puts CXLib.to_xml(read('books.cx'))

section 'books.cx  →  JSON  (repeated elements auto-collect into arrays)'
puts CXLib.to_json(read('books.cx'))

# ── cross-format round-trips ──────────────────────────────────────────────────

section 'books.xml   →  CX'
puts CXLib.xml_to_cx(read('books.xml'))

section 'books.json  →  CX'
puts CXLib.json_to_cx(read('books.json'))

section 'config.yaml  →  CX'
puts CXLib.yaml_to_cx(read('config.yaml'))

section 'config.toml  →  CX'
puts CXLib.toml_to_cx(read('config.toml'))

# ── vcore.cx: v3.3 features ───────────────────────────────────────────────────

section 'vcore.cx  (source)'
puts read('vcore.cx')

section 'vcore.cx  →  CX  (short aliases expand; auto-array gains explicit :type[])'
puts CXLib.to_cx(read('vcore.cx'))

section 'vcore.cx  →  JSON  (auto-array, :[], triple-quoted, block content)'
puts CXLib.to_json(read('vcore.cx'))

section 'vcore.cx  →  XML  (cx:type annotations; cx:block for block content)'
puts CXLib.to_xml(read('vcore.cx'))

# ── doc.cx: MD dialect document ───────────────────────────────────────────────

section 'doc.cx  (source)'
puts read('doc.cx')

section 'doc.cx  →  Markdown'
puts CXLib.to_md(read('doc.cx'))

section 'doc.cx  →  XML'
puts CXLib.to_xml(read('doc.cx'))

# ── doc.md: Markdown → CX ────────────────────────────────────────────────────

section 'doc.md  (source)'
puts read('doc.md')

section 'doc.md  →  CX'
puts CXLib.md_to_cx(read('doc.md'))

# ── AST inspection ────────────────────────────────────────────────────────────

# ── chapter.cx: XML-style structured document ─────────────────────────────────

section 'chapter.cx  (source)'
puts read('chapter.cx')

section 'chapter.cx  →  CX  (canonical)'
puts CXLib.to_cx(read('chapter.cx'))

section 'chapter.cx  →  XML  (structured document with sections and table)'
puts CXLib.to_xml(read('chapter.cx'))

section 'chapter.cx  →  JSON  (nested sections as nested objects)'
puts CXLib.to_json(read('chapter.cx'))

# ── post.cx: Markdown-style blog post ─────────────────────────────────────────

section 'post.cx  (source)'
puts read('post.cx')

section 'post.cx  →  Markdown  (CX markdown dialect → rendered Markdown)'
puts CXLib.to_md(read('post.cx'))

section 'post.cx  →  CX  (canonical)'
puts CXLib.to_cx(read('post.cx'))

section 'article.cx  →  AST  →  CX  (round-trip: AST JSON back to canonical CX)'
puts CXLib.ast_to_cx(CXLib.to_ast(read('article.cx')))

section 'article.cx  →  CX  (compact)'
puts CXLib.to_cx_compact(read('article.cx'))

# ── Document API ──────────────────────────────────────────────────────────────

section 'Document API: CXPath select'
src = "[services\n  [service name=auth  port=8080 active=true]\n  [service name=api   port=9000 active=false]\n  [service name=web   port=80   active=true]\n]"
doc = CXLib.parse(src)
first = doc.select('//service')
puts "first service: #{first.attr('name')}"
doc.select_all('//service[@active=true]').each { |svc| puts "active: #{svc.attr('name')}" }

section 'Document API: transform (immutable update)'
updated = doc.transform('services/service') { |el| el.set_attr('name', 'renamed-auth'); el }
puts updated.to_cx

section 'Document API: transform_all'
all_active = doc.transform_all('//service') { |el| el.set_attr('active', true); el }
puts "active services after transform_all: #{all_active.select_all('//service[@active=true]').length}"
